market_analysis/views/default.py

```diff
@@ -0,0 +1,450 @@
+from pyramid.response import Response
+from pyramid.view import view_config
+from pyramid.httpexceptions import HTTPFound
+from pyramid.security import remember, forget
+from sqlalchemy.exc import DBAPIError
+from sqlalchemy.orm.exc import UnmappedInstanceError
+from sqlalchemy import and_, or_
+try:
+    from urllib.parse import urlencode
+except ImportError:     # pragma: no cover
+    from urllib import urlencode
+from ..models import Stocks, Users, Association
+from ..security import check_credentials
+from passlib.apps import custom_app_context as pwd_context
+import datetime
+import requests
+
+
+@view_config(route_name='search', renderer='../templates/search.jinja2',
+                        permission='secret')
+def search_stocks(request):
+    msg = 'Search for stocks that you would like to watch.'
+    search_results = []
+    if request.method == 'GET':
+        return {'stocks': search_results, 'msg': msg}
+    elif request.method == 'POST':
+        try:
+            search = request.params.get('search')
+            search1 = search.lower().capitalize()
+            search2 = search.upper()
+            search_query = request.dbsession.query(Stocks)\
+                .filter(or_(Stocks.name.startswith(search1),
+                        Stocks.symbol.startswith(search2)))
+        except DBAPIError:  # pragma: no cover
+            return Response(db_err_msg, content_type='text/plain', status=500)
+
+        for row in search_query:
+            search_results.append(row)
+        if len(search_results) == 0:
+            msg = 'No results found, try again.'
+        return {'stocks': search_results, 'msg': msg}
+
+
+@view_config(route_name='add', renderer='../templates/add_page.jinja2',
+             permission='secret')
+def add_stock_to_portfolio(request):
+    if request.method == 'POST':
+        current_user_id = request.dbsession.query(Users).filter(
+                Users.username == request.authenticated_userid
+            ).first().id
+        new_stock_id = request.matchdict['id']
+        query = request.dbsession.query(Association)\
+            .filter(Association.user_id == current_user_id)
+        list_of_stock_ids = []
+        for row in query:
+            list_of_stock_ids.append(row.stock_id)
+        if int(new_stock_id) not in list_of_stock_ids:
+            association_row = Association(user_id=current_user_id,
+                                          stock_id=new_stock_id, shares=0)
+            request.dbsession.add(association_row)
+            msg = request.matchdict['name'] + ' was added to your portfolio.'
+        else:
+            msg = request.matchdict['name'] + ' is already in your portfolio.'
```

Comment:
It might be nice to make a method on the `Users` class that handles this.  Then you could do something like the following:

```python
current_user_query = request.dbsession.query(Users).filter(
    Users.username == request.authenticated_userid
)
current_user = current_user_query.first()
msg = current_user.maybe_add_stock_to_portfolio(request)
```

You could replace all of lines 48-63 with that.  Consider your models as a place to localize business logic.